crypto/aes/asm/aes-riscv64-zvkned.pl

```diff
@@ -0,0 +1,1413 @@
+    @{[vxor_vv $v24, $v24, $v16]}
+    j 2f
+
+1:
+    @{[vle32_v $v17, ($INP)]}
+    @{[vxor_vv $v24, $v24, $v17]}
+
+2:
+    # AES body
+    @{[aes_192_encrypt]}
+
+    @{[vse32_v $v24, ($OUTP)]}
+
+    addi $INP, $INP, 16
+    addi $OUTP, $OUTP, 16
+    addi $LEN, $LEN, -16
+
+    bnez $LEN, 1b
+
+    @{[vse32_v $v24, ($IVP)]}
+
+    ret
+.size L_cbc_enc_192,.-L_cbc_enc_192
+___
+
+$code .= <<___;
+.p2align 3
+L_cbc_enc_256:
+    # Load all 15 round keys to v1-v15 registers.
+    @{[aes_256_load_key]}
+
+    # Load IV.
+    @{[vle32_v $v16, ($IVP)]}
+
+    @{[vle32_v $v24, ($INP)]}
+    @{[vxor_vv $v24, $v24, $v16]}
+    j 2f
+
+1:
+    @{[vle32_v $v17, ($INP)]}
+    @{[vxor_vv $v24, $v24, $v17]}
+
+2:
+    # AES body
+    @{[aes_256_encrypt]}
+
+    @{[vse32_v $v24, ($OUTP)]}
+
+    addi $INP, $INP, 16
+    addi $OUTP, $OUTP, 16
+    addi $LEN, $LEN, -16
+
+    bnez $LEN, 1b
+
+    @{[vse32_v $v24, ($IVP)]}
+
+    ret
+.size L_cbc_enc_256,.-L_cbc_enc_256
+___
+
+###############################################################################
+# void rv64i_zvkned_cbc_decrypt(const unsigned char *in, unsigned char *out,
+#                               size_t length, const AES_KEY *key,
+#                               unsigned char *ivec, const int enc);
+my ($INP, $OUTP, $LEN, $KEYP, $IVP, $ENC) = ("a0", "a1", "a2", "a3", "a4", "a5");
+my ($T0, $T1, $rounds) = ("t0", "t1", "t2");
+my ($v0, $v1, $v2, $v3, $v4, $v5, $v6, $v7,
+    $v8, $v9, $v10, $v11, $v12, $v13, $v14, $v15,
+    $v16, $v17, $v18, $v19, $v20, $v21, $v22, $v23,
+    $v24, $v25, $v26, $v27, $v28, $v29, $v30, $v31,
+) = map("v$_",(0..31));
+
+$code .= <<___;
+.p2align 3
+.globl rv64i_zvkned_cbc_decrypt
+.type rv64i_zvkned_cbc_decrypt,\@function
+rv64i_zvkned_cbc_decrypt:
+    # check whether the length is a multiple of 16 and >= 16
+    li $T1, 16
+    blt $LEN, $T1, L_end
+    andi $T1, $LEN, 15
+    bnez $T1, L_end
+
+    # Load number of rounds
+    lwu $rounds, 240($KEYP)
+
+    # Get proper routine for key size
+    li $T0, 10
+    beq $rounds, $T0, L_cbc_dec_128
+
+    li $T0, 12
+    beq $rounds, $T0, L_cbc_dec_192
+
+    li $T0, 14
+    beq $rounds, $T0, L_cbc_dec_256
+
+    ret
+.size rv64i_zvkned_cbc_decrypt,.-rv64i_zvkned_cbc_decrypt
+___
+
+$code .= <<___;
+.p2align 3
+L_cbc_dec_128:
+    # Load all 11 round keys to v1-v11 registers.
+    @{[aes_128_load_key]}
+
+    # Load IV.
+    @{[vle32_v $v16, ($IVP)]}
+
+    @{[vle32_v $v24, ($INP)]}
+    @{[vmv_v_v $v17, $v24]}
+    j 2f
+
+1:
+    @{[vle32_v $v24, ($INP)]}
+    @{[vmv_v_v $v17, $v24]}
+    addi $OUTP, $OUTP, 16
+
+2:
+    # AES body
+    @{[aes_128_decrypt]}
+
+    @{[vxor_vv $v24, $v24, $v16]}
+    @{[vse32_v $v24, ($OUTP)]}
+    @{[vmv_v_v $v16, $v17]}
+
+    addi $LEN, $LEN, -16
+    addi $INP, $INP, 16
+
+    bnez $LEN, 1b
+
+    @{[vse32_v $v16, ($IVP)]}
+
+    ret
+.size L_cbc_dec_128,.-L_cbc_dec_128
+___
+
+$code .= <<___;
+.p2align 3
+L_cbc_dec_192:
+    # Load all 13 round keys to v10-v22 registers.
```

Comment:
How are the registers `v10-v22` selected ? (the default for `aes_192_load_key` seems to be `v1-v13`) ?

Reply:
and
```
my ($v0, $v1, $v2, $v3, $v4, $v5, $v6, $v7,
    $v8, $v9, $v10, $v11, $v12, $v13, $v14, $v15,
    $v16, $v17, $v18, $v19, $v20, $v21, $v22, $v23,
    $v24, $v25, $v26, $v27, $v28, $v29, $v30, $v31,
) = map("v$_",(0..31));
```
did not seem to contradict this.

Reply:
Fixed.